basex-core/src/main/java/org/basex/query/func/FuncRef.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * @author BaseX Team, BSD License
+ * @author Gunther Rademacher
+ */
+public class FuncRef extends Arr {
+  /** Function to resolve this reference. */
+  private final ThrowingFunction<QueryContext, Expr, QueryException> resolve;
+  /** Function to convert to string. */
+  private final Consumer<QueryString> toString;
+  /** Function call or function item after resolution. */
+  private Expr expr;
+
+  /**
+   * Constructor for static function calls.
+   * @param name function name
+   * @param fb function builder
+   * @param hasImport indicates whether a module import for the function name's URI was present
+   */
+  public FuncRef(final QNm name, final FuncBuilder fb, final boolean hasImport) {
+    this(fb.info,
+        qc -> Functions.get(name, fb, qc, hasImport),
+        qs -> qs.token(name.prefixId()).params(fb.args()));
+  }
+
+  /**
+   * Constructor for named function references.
+   * @param name function name
+   * @param arity function arity
+   * @param info input info (can be {@code null})
+   * @param hasImport indicates whether a module import for the function name's URI was present
+   */
+  public FuncRef(final QNm name, final int arity, final InputInfo info, final boolean hasImport) {
+    this(info,
+        qc -> Functions.item(name, arity, false, info, qc, hasImport),
+        qs -> qs.token(name.prefixId()).token('#').token(arity));
+  }
+
+  /**
+   * Constructor.
+   * @param info input info (can be {@code null})
+   * @param resolve function to resolve the reference
+   * @param toString function to convert to string
+   */
+  private FuncRef(final InputInfo info,
+      final ThrowingFunction<QueryContext, Expr, QueryException> resolve,
+      final Consumer<QueryString> toString) {
+    super(info, SeqType.ITEM_ZM);
+    this.resolve = resolve;
+    this.toString = toString;
+  }
+
+  /**
+   * Functional interface for functions that may throw an exception.
+   * @param <T> parameter type
+   * @param <R> result type
+   * @param <E> exception type
+   */
+  @FunctionalInterface
+  public interface ThrowingFunction<T, R, E extends Exception> {
+    /**
+     * Applies this function to the given argument.
+     * @param t argument
+     * @return result
+     * @throws E exception
+     */
+    R apply(T t) throws E;
+  }
+
+  /**
+   * Resolves the function reference.
+   * @param qc query context
+   * @throws QueryException query exception
+   */
+  public void resolve(final QueryContext qc) throws QueryException {
+    expr = resolve.apply(qc);
+  }
+
+  /**
+   * Returns the resolved expression.
+   * @return expression
+   */
+  private Expr expr() {
+    if(expr == null) throw Util.notExpected();
+    return expr;
+  }
+
+  @Override
+  public boolean has(final Flag... flags) {
+    return expr().has(flags);
+  }
+
+  @Override
+  public boolean vacuous() {
+    return expr().vacuous();
+  }
+
+  @Override
+  public Iter iter(final QueryContext qc) throws QueryException {
```

Comment:
We should try to get rid of all runtime references to `FuncRef`. For example, in the following expression that’s not the case yet:

```xquery
copy $c := parse-xml(<?_ <x/>?>)
modify ()
return $c
```

This one works: `parse-xml(<?_ <x/>?>) update {}`.